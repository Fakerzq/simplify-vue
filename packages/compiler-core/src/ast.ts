/*
 * @Author: Zhouqi
 * @Date: 2022-04-07 22:07:33
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-30 20:16:18
 */
import { CREATE_ELEMENT_VNODE, WITH_DIRECTIVES } from "./runtimeHelpers";

export const enum NodeTypes {
  ELEMENT,
  SIMPLE_EXPRESSION,
  INTERPOLATION,
  TEXT,
  ROOT,
  // containers
  COMPOUND_EXPRESSION,
  ATTRIBUTE,
  DIRECTIVE,
  COMMENT,
  VNODE_CALL,
  TEXT_CALL,
  JS_CALL_EXPRESSION,
  JS_PROPERTY,
  JS_OBJECT_EXPRESSION,
  JS_ARRAY_EXPRESSION,
  IF_BRANCH,
  IF,
  JS_CONDITIONAL_EXPRESSION,
}

export const enum ElementTypes {
  ELEMENT,
}

// 普通元素
export function createVnodeCall(context, tag, props, children, directives?) {
  context.helper(CREATE_ELEMENT_VNODE);
  if (directives) {
    context.helper(WITH_DIRECTIVES);
  }
  return {
    type: NodeTypes.VNODE_CALL,
    tag,
    props,
    children,
    directives,
  };
}

// 插值
export function createCallExpression(callee, args) {
  return {
    type: NodeTypes.JS_CALL_EXPRESSION,
    callee,
    arguments: args,
  };
}

// 指令 例如:v-show
export function createArrayExpression(elements) {
  return {
    type: NodeTypes.JS_ARRAY_EXPRESSION,
    elements,
  };
}

export function createConditionalExpression(test, consequent, alternate) {
  return {
    type: NodeTypes.JS_CONDITIONAL_EXPRESSION,
    test,
    consequent,
    alternate,
  };
}
