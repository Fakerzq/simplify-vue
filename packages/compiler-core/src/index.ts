/*
 * @Author: Zhouqi
 * @Date: 2022-04-10 14:22:24
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-05-10 21:00:32
 */
export * from "./compile";
export { registerRuntimeHelpers } from "./runtimeHelpers";
export { transformModel } from "./transforms/vModel";
export { NodeTypes } from "./ast";
