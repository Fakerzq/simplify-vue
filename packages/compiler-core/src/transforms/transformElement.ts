/*
 * @Author: Zhouqi
 * @Date: 2022-04-10 10:16:09
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-05-12 21:00:01
 */
import { isString, isSymbol, PatchFlags } from "@simplify-vue/shared";
import {
  createArrayExpression,
  createCallExpression,
  createSimpleExpression,
  createVnodeCall,
  ElementTypes,
  NodeTypes,
} from "../ast";
import { NORMALIZE_CLASS, RESOLVE_COMPONENT } from "../runtimeHelpers";
import { isStaticExp } from "../utils";

const directiveImportMap = new WeakMap();

export function transformElement(node, context) {
  return () => {
    const { type, tagType } = node;
    if (
      !(
        type === NodeTypes.ELEMENT &&
        (tagType === ElementTypes.ELEMENT || tagType === ElementTypes.COMPONENT)
      )
    )
      return;

    const { tag, props, children } = node;
    const isComponent = tagType === ElementTypes.COMPONENT;

    const vnodeTag = isComponent
      ? resolveComponentType(node, context)
      : `"${tag}"`;
    let vnodeProps;
    let vnodeChildren;
    let vnodeDirectives;
    let patchFlag = 0;
    let vnodePatchFlag;
    let dynamicPropNames;
    let vnodeDynamicProps;
    let shouldUseBlock = false;

    // 处理props
    if (props.length) {
      const buildResult = buildProps(node, context);
      vnodeProps = buildResult.props;
      patchFlag = buildResult.patchFlag;
      dynamicPropNames = buildResult.dynamicPropNames;
      const directives = buildResult.directives;
      // 处理指令
      if (directives && directives.length) {
        const buildDirectivesResult = directives.map((dir) =>
          buildDirectiveArgs(dir, context)
        );
        vnodeDirectives = createArrayExpression(buildDirectivesResult);
      }
    }

    if (children.length === 1) {
      const child = children[0];
      const type = child.type;
      // 是是否含有动态文本子节点
      // 1、插值节点
      // 2、 复合节点（文本+插值）
      const hasDynamicTextChild =
        type === NodeTypes.INTERPOLATION ||
        type === NodeTypes.COMPOUND_EXPRESSION;

      // 如果是动态文本节点，则标记patchFlag为TEXT
      if (hasDynamicTextChild) {
        patchFlag |= PatchFlags.TEXT;
      }

      if (hasDynamicTextChild || child.type === NodeTypes.TEXT) {
        // 如果只有一个子节点且是文本节点或者插值节点，则vnodeChildren为当前文本节点对象
        vnodeChildren = child;
      } else {
        // 否则为整个children
        vnodeChildren = children;
      }
    } else {
      // 孩子节点有多个的情况下vnodeChildren为children
      vnodeChildren = children;
    }

    if (patchFlag !== 0) {
      vnodePatchFlag = String(patchFlag);
      if (dynamicPropNames && dynamicPropNames.length) {
        vnodeDynamicProps = JSON.stringify(dynamicPropNames);
      }
    }
    // TODO shoulUseBlock
    node.codegenNode = createVnodeCall(
      context,
      vnodeTag,
      vnodeProps,
      vnodeChildren,
      vnodePatchFlag,
      vnodeDynamicProps,
      vnodeDirectives,
      shouldUseBlock,
      false,
      isComponent
    );
  };
}

/**
 * @author: Zhouqi
 * @description: 转化props
 * @param node
 * @param context
 * @return 转化后的props ast
 */
function buildProps(node, context) {
  const { props } = node;
  const properties: any = [];
  // 存储动态属性的数组
  const dynamicPropNames: string[] = [];
  let propsExpression;
  let runtimeDirectives: any = [];
  let patchFlag = 0;
  // 是否有v-bind的class
  let hasClassBinding = false;

  const analyzePatchFlag = (prop) => {
    const { key } = prop;
    // 处理静态的属性名
    if (isStaticExp(key)) {
      const name = key.content;

      if (name === "class") {
        hasClassBinding = true;
      } else if (!dynamicPropNames.includes(name)) {
        dynamicPropNames.push(name);
      }
    } else {
      // TODO处理动态的属性名
    }
  };

  for (let i = 0; i < props.length; i++) {
    const prop = props[i];
    // 处理静态属性
    if (prop.type === NodeTypes.ATTRIBUTE) {
      const { name, value } = prop;
      let isStatic = true;
      // 静态属性和值需要进行处理，处理对应的js ast表达式
      const nameExpression = createSimpleExpression(name, true);
      const valueExpression = createSimpleExpression(value.content, isStatic);
      properties.push(createObjectProperty(nameExpression, valueExpression));
    } else {
      // 处理指令
      const { name } = prop;
      // 处理v-bind、v-on、v-show
      const directiveTransform = context.directiveTransforms[name];
      if (directiveTransform) {
        const { props, needRuntime } = directiveTransform(prop, node, context);
        // 分析props上的属性，对不同性质的props进行动态信息的标记
        props.forEach(analyzePatchFlag);
        properties.push(...props);
        if (needRuntime) {
          runtimeDirectives.push(prop);
          // v-show
          if (isSymbol(needRuntime)) {
            directiveImportMap.set(prop, needRuntime);
          }
        }
      }
    }
  }

  if (properties.length) {
    propsExpression = createObjectExpression(dedupeProperties(properties));
  }

  if (propsExpression) {
    switch (propsExpression.type) {
      case NodeTypes.JS_OBJECT_EXPRESSION:
        // 找到props中的class属性ast，对于静态/动态的class需要额外处理
        const { properties } = propsExpression;
        let classPropIndex = -1;
        for (let i = 0; i < properties.length; i++) {
          const property = properties[i];
          if (property.key.content === "class") {
            classPropIndex = i;
          }
        }
        const classProp = properties[classPropIndex];
        if (classProp) {
          // 判断class属性值是不是动态的，如果是动态的需要添加辅助函数处理（normalizeClass）
          if (!isStaticExp(classProp.value)) {
            classProp.value = createCallExpression(
              context.helper(NORMALIZE_CLASS),
              [classProp.value]
            );
          }
        }
    }
  }

  // 如果有动态绑定的class，则标记为具有动态class的patchFlag
  if (hasClassBinding) {
    patchFlag |= PatchFlags.CLASS;
  }

  // 出了class和style，有其他动态属性
  if (dynamicPropNames.length) {
    patchFlag |= PatchFlags.PROPS;
  }

  // 有运行时指令的情况下需要添加NEED_PATCH标记，例如v-show
  if (patchFlag === 0 && runtimeDirectives.length > 0) {
    patchFlag |= PatchFlags.NEED_PATCH;
  }

  return {
    props: propsExpression,
    directives: runtimeDirectives,
    patchFlag,
    dynamicPropNames,
  };
}

export function createObjectProperty(key, value) {
  return {
    type: NodeTypes.JS_PROPERTY,
    key: isString(key) ? createSimpleExpression(key, true) : key,
    value,
  };
}

export function createObjectExpression(properties) {
  // JS_OBJECT_EXPRESSION 表示数据需要处理对象的形式，例如props
  return {
    type: NodeTypes.JS_OBJECT_EXPRESSION,
    properties,
  };
}

/**
 * @author: Zhouqi
 * @description: 去除重复属性
 * @param properties
 * @return 去重后的属性
 */
function dedupeProperties(properties) {
  let deduped: any = [];
  const kownProps = new Map();
  for (let i = 0; i < properties.length; i++) {
    const prop = properties[i];
    const name = prop.key.content;
    if (kownProps.has(name)) {
      // TODO
      // 对于class、style、事件需要进行值的追加
      // 例如：class="a" class="b" ===> class="a b"
    } else {
      // 如果没有重复的属性，则添加到去重后的结果中
      kownProps.set(name, prop);
      deduped.push(prop);
    }
  }
  return deduped;
}

/**
 * @author: Zhouqi
 * @description: 创建指令数据
 * @param  dir
 */
function buildDirectiveArgs(dir, context) {
  let dirArgs: any = [];
  const runtime = directiveImportMap.get(dir);
  if (runtime) {
    dirArgs.push(context.helperString(runtime));
  }
  if (dir.exp) dirArgs.push(dir.exp);
  return createArrayExpression(dirArgs);
}

/**
 * @author: Zhouqi
 * @description: 解析组件
 * @param node
 * @param context
 * @return
 */
function resolveComponentType(node, context) {
  const { tag } = node;
  context.helper(RESOLVE_COMPONENT);
  context.components.add(tag);
  return `_component_${tag}`;
}
