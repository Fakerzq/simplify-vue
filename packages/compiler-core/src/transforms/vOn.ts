/*
 * @Author: Zhouqi
 * @Date: 2022-04-23 19:47:29
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-26 09:37:29
 */
import { toHandlerKey } from "@simplify-vue/shared";
import {
  createObjectProperty,
  createSimpleExpression,
} from "./transformElement";

export function transformOn(dir) {
  const { arg, exp } = dir;
  // @click ===> onClick
  let eventName = createSimpleExpression(toHandlerKey(arg.content), true);
  const ret = createObjectProperty(eventName, exp);
  return { props: [ret] };
}
