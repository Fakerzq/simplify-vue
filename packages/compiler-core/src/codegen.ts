/*
 * @Author: Zhouqi
 * @Date: 2022-04-09 21:13:43
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-26 09:37:36
 */
/**
 * 1. text
 * return function render(_ctx, _cache, $props, $setup, $data, $options) {
 *   return "Hello World";
 * }
 *
 * 2、interplation
 * const { toDisplayString: _toDisplayString } = Vue
 *
 * return function render(_ctx, _cache, $props, $setup, $data, $options) {
 *    return _toDisplayString(_ctx.hello)
 * }
 *
 * 3、element
 * return function render(_ctx, _cache, $props, $setup, $data, $options) {
 *    return (_openBlock(), _createElementBlock("div"))
 * }
 *
 * 4、上面三种结合
 * const { toDisplayString: _toDisplayString, openBlock: _openBlock, createElementBlock: _createElementBlock } = Vue
 *  return function render(_ctx, _cache, $props, $setup, $data, $options) {
 *    return (_openBlock(), _createElementBlock("div", null, "hello, " + _toDisplayString(_ctx.message), 1));
 *  }
 */

import { isArray, isFunction, isString } from "@simplify-vue/shared";
import { NodeTypes } from "./ast";
import {
  CREATE_ELEMENT_BLOCK,
  CREATE_ELEMENT_VNODE,
  helperNameMap,
  OPEN_BLOCK,
  TO_DISPLAY_STRING,
  WITH_DIRECTIVES,
} from "./runtimeHelpers";

export function generate(ast, options = {}) {
  const context = createCodegenContext(ast, options);

  // 生成前导：例如 const { toDisplayString: _toDisplayString, openBlock: _openBlock, createElementBlock: _createElementBlock } = Vue
  genFunctionPreamble(ast, context);

  const { push, newLine } = context;
  const functionName = "render";
  const args = ["_ctx", "_cache", "$props", "$setup", "$data", "$options"].join(
    ", "
  );
  push(`function ${functionName}(${args}) { `);
  newLine();
  push(`return `);
  genNode(ast.codegenNode, context);
  push(`}`);

  return { code: context.code };
}

/**
 * @author: Zhouqi
 * @description: 创建codegen上下文
 * @param ast
 * @param options
 * @return 上下文对象
 */
function createCodegenContext(ast: any, options: {}) {
  const context = {
    runtimeGlobalName: `Vue`,
    // 最终的代码字符串
    code: ``,
    // 字符串拼接操作
    push(text) {
      context.code += text;
    },
    newLine() {
      context.push(`\n      `);
    },
    helper(key) {
      return `_${helperNameMap[key]}`;
    },
  };
  return context;
}

/**
 * @author: Zhouqi
 * @description: 根据ast生成节点
 * @param node astNode
 */
function genNode(node, context) {
  const { type } = node;
  const nodeHandlers = {
    [(type === NodeTypes.TEXT) as any]() {
      // 文本
      genText(node, context);
    },
    [(type === NodeTypes.INTERPOLATION) as any]() {
      // 插值
      genInterpolation(node, context);
    },
    [(type === NodeTypes.SIMPLE_EXPRESSION) as any]() {
      // 值，例如属性名、属性值
      genExpression(node, context);
    },
    [(type === NodeTypes.ELEMENT) as any]() {
      // 元素
      genNode(node.codegenNode, context);
    },
    [(type === NodeTypes.COMPOUND_EXPRESSION) as any]() {
      // 复合类型 text+{{}}
      genCompoundExpression(node, context);
    },
    [(type === NodeTypes.VNODE_CALL) as any]() {
      genVNodeCall(node, context);
    },
    [(type === NodeTypes.TEXT_CALL) as any]() {
      // 节点
      genNode(node.codegenNode, context);
    },
    [(type === NodeTypes.JS_CALL_EXPRESSION) as any]() {
      genCallExpression(node, context);
    },
    [(type === NodeTypes.JS_OBJECT_EXPRESSION) as any]() {
      // 属性
      genObjectExpression(node, context);
    },
    [(type === NodeTypes.JS_ARRAY_EXPRESSION) as any]() {
      // 指令
      genArrayExpression(node, context);
    },
  };
  const handler = nodeHandlers[true as any];
  if (isFunction(handler)) {
    handler();
  }
}

function genArrayExpression(node, context) {
  genNodeListAsArray(node.elements, context);
}

function genObjectExpression(node, context) {
  const { push } = context;
  const { properties } = node;
  push(`{ `);
  for (let i = 0; i < properties.length; i++) {
    const { key, value } = properties[i];
    // 处理key
    genExpressionAsPropertyKey(key, context);
    push(`: `);
    // 处理value
    genNode(value, context);
    // {xx:xxx, saa:saa}
    if (i < properties.length - 1) {
      push(`,`);
    }
  }
  push(`} `);
}

/**
 * @author: Zhouqi
 * @description: 处理属性key
 * @param key
 * @param context
 */
function genExpressionAsPropertyKey(key, context) {
  const { push } = context;
  const text = key.content;
  push(text);
}

/**
 * @author: Zhouqi
 * @description: 处理多子节点中含有文本/插值表达式的情况
 * @param node
 * @param context
 */
function genCallExpression(node, context) {
  const { push, helper } = context;
  const callee = isString(node.callee) ? node.callee : helper(node.callee);
  push(callee + `(`);
  // 处理节点
  genNodeList(node.arguments, context);
  push(`)`);
}

// 处理虚拟节点
function genVNodeCall(node, context) {
  const { push, helper } = context;
  const { isBlock, tag, props, children, directives } = node;
  if (directives) {
    push(helper(WITH_DIRECTIVES) + `(`);
  }
  if (isBlock) {
    push(`(${helper(OPEN_BLOCK)}(${``}), `);
  }
  const callHelper = isBlock ? CREATE_ELEMENT_BLOCK : CREATE_ELEMENT_VNODE;
  push(helper(callHelper) + `(`);
  genNodeList(genNullableArgs([tag, props, children]), context);
  push(`)`);
  if (isBlock) {
    push(`)`);
  }
  if (directives) {
    push(`, `);
    genNode(directives, context);
    push(`)`);
  }
}

/**
 * @author: Zhouqi
 * @description: 生成前导函数代码
 * @param ast
 * @param context
 */
function genFunctionPreamble(ast, context) {
  const { push, runtimeGlobalName, newLine } = context;
  const { helpers } = ast;

  const aliasHelper = (s: symbol) =>
    `${helperNameMap[s]}: _${helperNameMap[s]}`;

  if (helpers.length) {
    push(
      `const { ${helpers.map(aliasHelper).join(", ")} } = ${runtimeGlobalName};`
    );
  }
  newLine();
  push(`return `);
}

/**
 * @author: Zhouqi
 * @description: 把不存在的属性都变成null
 * @param args 属性
 * @return 处理后的属性
 */
function genNullableArgs(args) {
  return args.map((arg) => arg || "null");
}

/**
 * @author: Zhouqi
 * @description: 处理孩子节点
 * @param nodes
 * @param context
 */
function genNodeListAsArray(nodes, context) {
  // 子节点需要加上[]包裹 h('xx',null, [children])
  context.push(`[`);
  genNodeList(nodes, context);
  context.push(`]`);
}

/**
 * @author: Zhouqi
 * @description: 处理列表形式的节点  ['p', props ,children]
 * @param nodes 列表形式的节点
 * @param context
 */
function genNodeList(nodes, context) {
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      push(node);
    } else if (isArray(node)) {
      // 处理孩子节点
      genNodeListAsArray(node, context);
    } else {
      genNode(node, context);
    }
    if (i < nodes.length - 1) {
      push(`, `);
    }
  }
}

/**
 * @author: Zhouqi
 * @description: 生成文本节点的代码字符串
 * @param node
 * @param context
 */
function genText(node, context) {
  context.push(JSON.stringify(node.content));
}

/**
 * @author: Zhouqi
 * @description: 处理插值节点
 * @param node
 * @param context
 */
function genInterpolation(node, context) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(`)`);
}

/**
 * @author: Zhouqi
 * @description: 生成插值中的表达式的代码字符串
 * @param node
 * @param context
 */
function genExpression(node, context) {
  const { isStatic, content } = node;
  // 静态静态节点转换成字符串
  context.push(isStatic ? JSON.stringify(content) : content);
}

/**
 * @author: Zhouqi
 * @description: 处理复合节点
 * @param node
 * @param context
 */
function genCompoundExpression(node, context) {
  const { children } = node;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (isString(child)) {
      // +
      context.push(child);
    } else {
      genNode(child, context);
    }
  }
}
