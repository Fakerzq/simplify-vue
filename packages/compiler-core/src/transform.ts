/*
 * @Author: Zhouqi
 * @Date: 2022-04-09 20:33:38
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-30 13:47:59
 */
import { isFunction, isArray } from "@simplify-vue/shared";
import { NodeTypes } from "./ast";
import {
  OPEN_BLOCK,
  TO_DISPLAY_STRING,
  CREATE_ELEMENT_BLOCK,
  helperNameMap,
} from "./runtimeHelpers";

/**
 * @author: Zhouqi
 * @description: 转化ast
 * @param root ast
 * @param options 配置
 * @return
 */
export function transform(root, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
  createRootCodegen(root, context);
  root.helpers = [...context.helpers.keys()];
}

/**
 * @author: Zhouqi
 * @description: 创建codegen所需要的ast
 * @param root ast
 */
function createRootCodegen(root, context) {
  const { children } = root;

  // root下只有一个子节点，即单根标签的情况
  if (children.length === 1) {
    const child = root.children[0];
    if (child.type === NodeTypes.ELEMENT && child.codegenNode) {
      const codegenNode = child.codegenNode;
      if (codegenNode.type === NodeTypes.VNODE_CALL) {
        makeBlock(codegenNode, context);
      }
      root.codegenNode = codegenNode;
    } else {
      root.codegenNode = child;
    }
  }
}

function makeBlock(node, context) {
  const { helper } = context;
  node.isBlock = true;
  helper(OPEN_BLOCK);
  helper(CREATE_ELEMENT_BLOCK);
}

/**
 * @author: Zhouqi
 * @description: 创换上下文对象
 * @param root ast
 * @param options 配置
 * @return 上下文对象
 */
function createTransformContext(
  root,
  { nodeTransforms = [], directiveTransforms = {} }
) {
  const context = {
    directiveTransforms,
    // 当前正在转换的节点
    currentNode: root,
    // 父节点
    parent: null,
    // 当前节点在父节点children中索引
    childIndex: 0,
    root,
    nodeTransforms,
    helpers: new Map(),
    helper(name) {
      context.helpers.set(name, 1);
      return name;
    },
    helperString(name) {
      return `_${helperNameMap[context.helper(name)]}`;
    },
    replaceNode(node) {
      // 替换当前节点的同时修改父节点上对应的子节点
      (context.parent! as any).children[context.childIndex] =
        context.currentNode = node;
    },
    removeNode(node) {
      const children = (context.parent as any).children;
      const nodeIndex = node
        ? children.indexOf(node)
        : context.currentNode
        ? context.childIndex
        : -1;
      // 上下文中清空当前节点
      context.currentNode = null;
      // 此时在遍历traverseChildren的时候，由于当前node会从children中删除，会影响到遍历
      // 因此需要调用onNodeRemoved方法，将正在遍历的index--
      context.onNodeRemoved();
      children.splice(nodeIndex, 1);
    },
    onNodeRemoved: () => {},
  };

  return context;
}
/**
 * @author: Zhouqi
 * @description: 转化ast节点
 * @param node ast
 * @param context 上下文对象
 * @return
 */
export function traverseNode(node, context) {
  // 设置当前正在转换的节点
  context.currentNode = node;
  const { nodeTransforms } = context;

  //   if (type === NodeTypes.TEXT) {
  //     node.content = node.content + "123";
  //   }
  // 插件模式和洋葱模型
  const exitFns: any = [];
  for (let i = 0; i < nodeTransforms.length; i++) {
    const onExit = nodeTransforms[i](node, context);

    if (onExit) {
      // v-if
      if (isArray(onExit)) {
        exitFns.push(...onExit);
      } else {
        exitFns.push(onExit);
      }
    }

    if (!context.currentNode) {
      // 说明节点被删除了，此时直接返回即可，不进行接下去的处理
      return;
    } else {
      // 当前节点可能会被替换，所以要重新赋值一次，例如v-if处理元素时，当前ast节点会被替换成指定特殊的ast结构
      node = context.currentNode;
    }
  }

  const { type } = node;

  switch (type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING);
      break;
    case NodeTypes.IF:
      for (let i = 0; i < node.branches.length; i++) {
        traverseNode(node.branches[i], context);
      }
      break;
    case NodeTypes.ELEMENT:
    case NodeTypes.ROOT:
    case NodeTypes.IF_BRANCH:
      traverseChildren(node, context);
      break;
  }

  context.currentNode = node;
  let i = exitFns.length;
  while (i--) {
    exitFns[i]();
  }
}

/**
 * @author: Zhouqi
 * @description: ast子节点转化
 * @param parent 父节点
 * @param context 上下文对象
 * @return
 */
function traverseChildren(parent, context) {
  // 遍历子节点，递归处理
  let i = 0;
  // 存在遍历当前节点，节点会被移除的情况，这里绑定一个nodeRemoved方法，
  // 当节点被删除时可以通过context触发这个函数，从而让遍历正常进行
  const nodeRemoved = () => {
    i--;
  };
  for (i; i < parent.children.length; i++) {
    const child = parent.children[i];
    context.parent = parent;
    context.childIndex = i;
    context.onNodeRemoved = nodeRemoved;
    traverseNode(child, context);
  }
}

// 结构化指令转换器（v-if/v-else-if/v-else）
export function createStructuralDirectiveTransform(name, fn) {
  return (node, context) => {
    if (node.type === NodeTypes.ELEMENT) {
      const props = node.props;
      const existFn: any = [];
      for (let i = 0; i < props.length; i++) {
        const prop = props[i];
        // 如果属性名匹配到了对应的结构化指令
        if (prop.type === NodeTypes.DIRECTIVE && name.test(prop.name)) {
          /**
           * 从当前element元素的props中删除匹配到的结构化指令数据，避免无限递归
           * 
           * 原因：
           * 因为这里接下去会走processIf流程，这个流程会创建if_branch，if_branch中
           * 的children存储着当前节点，当transverseNode处理if_branch的时候会transverseChildren，
           * 此时又会遍历到当前这个节点，由于当前这个节点的props依旧存在。所以还是会继续走这里的流程，
           * 从而形成无限递归的死循环现象。
           */
          props.splice(i, 1);
          i--;
          const onExist = fn(node, prop, context);
          onExist && existFn.push(onExist);
        }
      }
      return existFn;
    }
  };
}
