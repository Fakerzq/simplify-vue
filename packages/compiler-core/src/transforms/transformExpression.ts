/*
 * @Author: Zhouqi
 * @Date: 2022-04-09 23:38:10
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-23 11:27:05
 */
import { NodeTypes } from "../ast";

export function transformExpression(node) {
  if (node.type === NodeTypes.INTERPOLATION) {
    node.content = processExpression(node.content);
  } else if (node.type === NodeTypes.ELEMENT) {
    // 处理普通元素上的指令
    const { props } = node;
    for (let i = 0; i < props.length; i++) {
      const dir = props[i];
      if (dir.type === NodeTypes.DIRECTIVE) {
        const { exp } = dir;

        // 处理动态属性值
        if (exp && exp.type === NodeTypes.SIMPLE_EXPRESSION) {
          dir.exp = processExpression(exp);
        }
      }
    }
  }
}

function processExpression(node) {
  node.content = `_ctx.${node.content}`;
  return node;
}
