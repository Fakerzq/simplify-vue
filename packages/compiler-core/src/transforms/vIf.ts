/*
 * @Author: Zhouqi
 * @Date: 2022-04-26 21:19:36
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-27 20:49:17
 */
import {
  createCallExpression,
  createConditionalExpression,
  NodeTypes,
} from "../ast";
import { CREATE_COMMENT } from "../runtimeHelpers";
import { createStructuralDirectiveTransform } from "../transform";
import { processExpression } from "./transformExpression";

// v-if指令转换器
export const transformIf = createStructuralDirectiveTransform(
  /^(if|else|else-if)$/,
  (node, dir, context) => {
    return processIf(node, dir, context, (ifNode, branch, isRoot) => {
      return () => {
        if (isRoot) {
          // v-if
          ifNode.codegenNode = createCodegenNodeForBranch(branch, context);
        } else {
          // TODO else if / else
        }
      };
    });
  }
);

function processIf(node, dir, context, fn?) {
  const { name, exp } = dir;
  if (exp) {
    dir.exp = processExpression(exp);
  }

  if (name === "if") {
    // 处理 if
    const branch = createIfBranch(node, dir);
    const ifNode = {
      type: NodeTypes.IF,
      branches: [branch],
    };
    context.replaceNode(ifNode);
    if (fn) {
      return fn(ifNode, branch, true);
    }
  } else {
    // TODO else if / else
  }
}

function createIfBranch(node, dir) {
  return {
    type: NodeTypes.IF_BRANCH,
    condition: dir.exp,
    children: [node],
  };
}

function createCodegenNodeForBranch(branch, context) {
  if (branch.condition) {
    return createConditionalExpression(
      branch.condition,
      // 表达式是true是需要创建的元素ast
      createChildrenCodegenNode(branch, context),
      // 表达式是false时创建的注释节点ast
      createCallExpression(context.helper(CREATE_COMMENT), ['"v-if"', "true"])
    );
  }
}
function createChildrenCodegenNode(branch, context) {
  return branch.children[0];
}
