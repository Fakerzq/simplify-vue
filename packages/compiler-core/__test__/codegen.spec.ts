/*
 * @Author: Zhouqi
 * @Date: 2022-04-09 20:34:26
 * @LastEditors: Zhouqi
 * @LastEditTime: 2022-04-30 20:00:48
 */
import { transformShow } from "../../compiler-dom/src/transform/vShow";
import { generate } from "../src/codegen";
import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { transformElement } from "../src/transforms/transformElement";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformText } from "../src/transforms/transformText";
import { transformBind } from "../src/transforms/vBind";
import { transformIf } from "../src/transforms/vIf";
import { transformOn } from "../src/transforms/vOn";
describe("Compiler: transform", () => {
  test("context state", () => {
    const ast = baseParse(
      `<div>1</div><div>2</div>`
    );

    transform(ast, {
      nodeTransforms: [transformIf, transformExpression, transformElement],
      // nodeTransforms: [transformExpression, transformElement, transformText],
      // directiveTransforms: {
      //   bind: transformBind,
      //   on: transformOn,
      //   show: transformShow,
      // },
    });
    const code = generate(ast);
    expect(code).toMatchSnapshot();
  });
  // test("codegen string", () => {
  //   const ast = baseParse(`hello`);
  //   transform(ast);
  //   const { code } = generate(ast);

  //   expect(code).toMatchSnapshot();
  // });

  // test("codegen interplation", () => {
  //   const ast = baseParse(`{{ hello }}`);

  //   transform(ast, {
  //     nodeTransforms: [transformExpression],
  //   });
  //   const { code } = generate(ast);

  //   expect(code).toMatchSnapshot();
  // });

  // test("codegen element", () => {
  //   const ast = baseParse(`<div></div>`);

  //   transform(ast, {
  //     nodeTransforms: [transformElement],
  //   });
  //   const { code } = generate(ast);

  //   expect(code).toMatchSnapshot();
  // });

  // test("codegen element interplation string", () => {
  //   const ast = baseParse(`<div>hello, {{ message }}</div>`);

  //   transform(ast, {
  //     nodeTransforms: [transformElement, transformExpression, transformText],
  //   });

  //   const { code } = generate(ast);

  //   expect(code).toMatchSnapshot();
  // });
});
