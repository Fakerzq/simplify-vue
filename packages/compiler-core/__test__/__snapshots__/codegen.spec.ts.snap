// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler: transform context state 1`] = `
Object {
  "code": "const { vShow: _vShow, createElementVNode: _createElementVNode, withDirectives: _withDirectives, openBlock: _openBlock, createElementBlock: _createElementBlock } = Vue;
      return function render(_ctx, _cache, $props, $setup, $data, $options) { 
      return _withDirectives((_openBlock(), _createElementBlock(\\"div\\", null, [])), [[_vShow, _ctx.show]])}",
}
`;
